LV2 Worker feature

ctsexton opened the pull request:
Hey there, 

Thank you for this repo!

I am pretty new to both Rust and LV2, but this is my attempt to add support for the [LV2 Worker feature](http://lv2plug.in/ns/ext/worker). It's still a work in progress and not yet adequately tested (like, not at all) but I figured now would be a good time to open a draft PR and potentially get some feedback if possible as the basic structure is there.

Honestly this was pretty challenging and I had to use a lot of unsafe code that made me a bit uncomfortable. But I guess the worker feature _is_ unsafe by nature - running the plugin in the real-time thread while it simultaneously does work in another thread is pretty much the definition of unsafe.

Here is some pseudo-code of how I intended it to work:

```
let instance = plugin.instantiate(...)
let worker = instance.get_worker().unwrap()

// In real-time thread:
plugin.run()

// In another thread
while(plugin_is_alive) {
  worker.do_work()
}
```

I am not yet sure how to make sure the worker does not outlive the plugin instance - I have a bit to learn yet on rust lifetimes.

wmedrano commented:
If I understand correctly, `worker.do_work` should be run continuously as long as the plugin instance is alive. The plugin instance dies during `Instance::drop`. If that's the case, how about:

```rust
struct Instance {
    ...
    is_alive: std::sync::Arc<std::sync::Mutex<bool>>, // true while alive
}
let instance = plugin.instantiate(...)
struct Worker {
    ...
    is_alive: std::sync::Arc<std::sync::Mutex<bool>>, // shared with instance
}
let worker = instance.get_worker().unwrap()

// In real-time thread:
plugin.run()

// In thread where items are dropped. Hopefully not realtime, but things happen lol.
// In the drop method:
//    1. is_alive mutex will attempt to lock. May have to wait if `worker.do_work` is running.
//    2. is_alive is set to false. `worker.do_work` should no longer do anything.
//    3. the underlying `lilv::instance::ActiveInstance` drop methods are run and `worker.do_work` should no longer run.
drop(plugin)

// In another thread.
let is_alive = worker.is_alive.lock().unwrap();
while(is_alive) {
  worker.do_work()
}
drop(is_alive); // Release the mutex. This happens implicitly so no need to manual drop.
```

ctsexton commented:
Thanks for the mutex suggestion, that should work well - I forgot we could lock a mutex in the drop method. I have implemented it pretty much as you described. I've also added a `WorkerManager` to run workers and drop them when their respective plugin is no longer alive.

So, intended usage is now:
```
let worker_manager = WorkerManager::new();

let instance = plugin.instantiate(...);

if let Some(worker) = instance.get_worker() {
  worker_manager.add_worker(worker);
}

// real-time thread
instance.run()

// non-real-time thread/main loop
while(true) {
  worker_manager.run_workers();
  // wait a short time or do something else
}
```

What else would you need done to get this into a merge-able state? 

I would ideally want to have some unit tests that demonstrate this working with a plugin that uses the worker feature. One such plugin is the setBfree organ. I don't love the idea of having to install third party plugins just to make unit tests pass, but I guess that is a lot less work than writing plugin code into this library as well. I did some reading on mocking in rust to see if it was possible to mock out lilv library components but it seems like we'd have to substantially change the plugin.rs code to hide the ActiveInstance and lilv::plugin::Plugin behind traits. What do you think?

wmedrano commented:
Testing a plugin is the safest thing. For example, src/lib.rs has test_mda_epiano. You can do something similar with setBfree organ. For GitHub actions, you can add `setbfree` to `.github/workflows/test.yml` line 19.

Eventually, I'd like to test most plugins anyways.

wmedrano commented:
I think the lint error can be resolved with adding the following to the affected structs/functions
`#[allow(clippy::mutex_atomic)]` We definitely don't want `AtomicBool` as holding the lock is desired.

wmedrano commented on src/features/mod.rs:
```diff
@@ -5,6 +5,7 @@ use std::{collections::HashSet, ffi::CStr};
 
 pub mod options;
 pub mod urid_map;
+pub mod worker;
```
`iter_features` from this file also needs to include the worker feature for the plugin to not get filtered out.

Some solutions I see:
- Remove `iter_features` and inline all uses. I think its mostly for `supported_features` (in this file) and building the features when instantiating plugins.
- Have `iter_features` take in the worker `LV2Feature` as a param. For `supported_features`, you can pass in a constant that just has the URI with data set to `std::ptr::null`. See `BOUNDED_BLOCK_LENGTH` defined in this file.

ctsexton replied:
Thanks, I took the first approach. After thinking about it a bit I figured that the HashSet returned by `supported_features` and iterating over the features themselves are quite different use cases, so there's no need for `supported_features` to use on `iter_features`. I've simplified `supported_features` to just return `HashSet<&'static str>` since the contents of that I assume should not be expected to change at runtime?

wmedrano reviewed (commented):
Also export `Worker` and `WorkerManager` in `src/lib.rs`

```rust
pub use features::worker::{Worker, WorkerManager};
```

ctsexton commented:
Have updated per your feedback! It looks to be in reasonable shape now I think except I still need to add a test. I'll try to work on writing a test that uses setBFree or similar to schedule and perform some work, I guess all in a single thread and calling `worker.do_work()` between calls to `instance.run()`. 

I agree it's nice to test real plugins, and ideally many, but that is a lot of work. I wonder if there is a plugin out there written for test purposes that implements as many features as possible, or perhaps that would be a useful thing to develop in parallel with a library like livi. The other thing I noticed is each test that loads a plugin takes a while to set up because lilv has to initialize all the plugins, which will get out of hand pretty quickly as you add more tests. I think that would be avoided by adding support for [`load_bundle`](https://docs.rs/lilv/latest/lilv/struct.World.html#method.load_bundle) though, so perhaps I can try that.

wmedrano commented:
I tried running your fork with `http://gareus.org/oss/lv2/b_synth` and there seem to be issues with the way livi is handling midi/events.

Good news is I managed to fix the issues (by referencing [jalv](https://github.com/drobilla/jalv)) and will push out a fix this weekend. Other than that, this PR seems to work!

wmedrano commented:
WIP: https://github.com/wmedrano/livi-rs/pull/13

ctsexton commented:
Thanks, I have rebased and written an integration test. I ended up using eg-sampler instead of b_synth because it was much easier to figure out how to build a message to trigger some work to be done. The test file is hopefully self-explanatory, but in summary:
- we write a sample to a file
- we load the sample in the sampler plugin (using the worker feature)
- we play the sample to prove it loaded correctly

For some reason the tests are failing on github. I'm not sure what is happening there. I tried scoping the tests command in the workflow file to only run my new test (`cargo test --verbose test_sampler -- --nocapture`) and it ran fine, so perhaps another test is failing?

ctsexton commented:
There is a seg fault when getting the worker interface from the eg-metro plugin which has extension_data = NULL. I _think_ this could be a bug with the lv2_raw lib so I opened an issue there: https://github.com/poidl/lv2_raw/issues/4

I'll try to come up with a workaround to check if extension_data is null but not sure how to do that!

wmedrano commented:
Looks like this can be fixed in lilv_sys.
```
wmedrano@donuts:~/Git/ctsexton/livi-rs$ valgrind /home/wmedrano/Git/ctsexton/livi-rs/target/debug/deps/livi-d656d3ff69ab653a test_all_plugins --nocapture
==25281== Memcheck, a memory error detector
==25281== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==25281== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==25281== Command: /home/wmedrano/Git/ctsexton/livi-rs/target/debug/deps/livi-d656d3ff69ab653a test_all_plugins --nocapture
==25281== 
==25281== Syscall param statx(file_name) points to unaddressable byte(s)
==25281==    at 0x4B479FE: statx (statx.c:29)
==25281==    by 0x204B50: statx (weak.rs:139)
==25281==    by 0x204B50: std::sys::unix::fs::try_statx (fs.rs:139)
==25281==    by 0x2013A2: std::sys::unix::fs::stat (fs.rs:1188)
==25281==    by 0x1C5A54: metadata<&std::path::PathBuf> (fs.rs:1636)
==25281==    by 0x1C5A54: get_dbpath_for_term (searcher.rs:50)
==25281==    by 0x1C5A54: from_name (mod.rs:83)
==25281==    by 0x1C5A54: from_env (mod.rs:69)
==25281==    by 0x1C5A54: new<std::io::stdio::Stdout> (mod.rs:159)
==25281==    by 0x1C5A54: test::term::stdout (term.rs:33)
==25281==    by 0x1B1D55: test::console::run_tests_console (console.rs:255)
==25281==    by 0x1C7B64: test::test_main (lib.rs:124)
==25281==    by 0x1C8C70: test::test_main_static (lib.rs:143)
==25281==    by 0x12FFA2: livi::main (lib.rs:1)
==25281==    by 0x139A0A: core::ops::function::FnOnce::call_once (function.rs:227)
==25281==    by 0x18323D: std::sys_common::backtrace::__rust_begin_short_backtrace (backtrace.rs:123)
==25281==    by 0x14ABB0: std::rt::lang_start::{{closure}} (rt.rs:145)
==25281==    by 0x1FD3CA: call_once<(), (dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe)> (function.rs:259)
==25281==    by 0x1FD3CA: do_call<&(dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe), i32> (panicking.rs:406)
==25281==    by 0x1FD3CA: try<i32, &(dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe)> (panicking.rs:370)
==25281==    by 0x1FD3CA: catch_unwind<&(dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe), i32> (panic.rs:133)
==25281==    by 0x1FD3CA: {closure#2} (rt.rs:128)
==25281==    by 0x1FD3CA: do_call<std::rt::lang_start_internal::{closure#2}, isize> (panicking.rs:406)
==25281==    by 0x1FD3CA: try<isize, std::rt::lang_start_internal::{closure#2}> (panicking.rs:370)
==25281==    by 0x1FD3CA: catch_unwind<std::rt::lang_start_internal::{closure#2}, isize> (panic.rs:133)
==25281==    by 0x1FD3CA: std::rt::lang_start_internal (rt.rs:128)
==25281==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==25281== 
==25281== Syscall param statx(buf) points to unaddressable byte(s)
==25281==    at 0x4B479FE: statx (statx.c:29)
==25281==    by 0x204B50: statx (weak.rs:139)
==25281==    by 0x204B50: std::sys::unix::fs::try_statx (fs.rs:139)
==25281==    by 0x2013A2: std::sys::unix::fs::stat (fs.rs:1188)
==25281==    by 0x1C5A54: metadata<&std::path::PathBuf> (fs.rs:1636)
==25281==    by 0x1C5A54: get_dbpath_for_term (searcher.rs:50)
==25281==    by 0x1C5A54: from_name (mod.rs:83)
==25281==    by 0x1C5A54: from_env (mod.rs:69)
==25281==    by 0x1C5A54: new<std::io::stdio::Stdout> (mod.rs:159)
==25281==    by 0x1C5A54: test::term::stdout (term.rs:33)
==25281==    by 0x1B1D55: test::console::run_tests_console (console.rs:255)
==25281==    by 0x1C7B64: test::test_main (lib.rs:124)
==25281==    by 0x1C8C70: test::test_main_static (lib.rs:143)
==25281==    by 0x12FFA2: livi::main (lib.rs:1)
==25281==    by 0x139A0A: core::ops::function::FnOnce::call_once (function.rs:227)
==25281==    by 0x18323D: std::sys_common::backtrace::__rust_begin_short_backtrace (backtrace.rs:123)
==25281==    by 0x14ABB0: std::rt::lang_start::{{closure}} (rt.rs:145)
==25281==    by 0x1FD3CA: call_once<(), (dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe)> (function.rs:259)
==25281==    by 0x1FD3CA: do_call<&(dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe), i32> (panicking.rs:406)
==25281==    by 0x1FD3CA: try<i32, &(dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe)> (panicking.rs:370)
==25281==    by 0x1FD3CA: catch_unwind<&(dyn core::ops::function::Fn<(), Output=i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe), i32> (panic.rs:133)
==25281==    by 0x1FD3CA: {closure#2} (rt.rs:128)
==25281==    by 0x1FD3CA: do_call<std::rt::lang_start_internal::{closure#2}, isize> (panicking.rs:406)
==25281==    by 0x1FD3CA: try<isize, std::rt::lang_start_internal::{closure#2}> (panicking.rs:370)
==25281==    by 0x1FD3CA: catch_unwind<std::rt::lang_start_internal::{closure#2}, isize> (panic.rs:133)
==25281==    by 0x1FD3CA: std::rt::lang_start_internal (rt.rs:128)
==25281==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==25281== 

running 1 test
Testing block_size 64 with plugin: http://lv2plug.in/plugins/eg-metro
Get instance
Get extension data
==25281== Thread 2 tests::test_all_:
==25281== Jump to the invalid address stated on the next line
==25281==    at 0x0: ???
==25281==    by 0x14AE62: lilv_instance_get_extension_data (inline_fns.rs:106)
==25281==    by 0x14AE62: lilv::instance::Instance::extension_data (instance.rs:115)
==25281==    by 0x16EFD5: livi::features::worker::maybe_get_worker_interface (worker.rs:153)
==25281==    by 0x17986A: livi::plugin::Plugin::instantiate (plugin.rs:103)
==25281==    by 0x15439D: livi::tests::test_all_plugins (lib.rs:338)
==25281==    by 0x17B4B9: livi::tests::test_all_plugins::{{closure}} (lib.rs:291)
==25281==    by 0x13995D: core::ops::function::FnOnce::call_once (function.rs:227)
==25281==    by 0x1D2D72: call_once<fn(), ()> (function.rs:227)
==25281==    by 0x1D2D72: test::__rust_begin_short_backtrace (lib.rs:585)
==25281==    by 0x1D17F3: call_once<(), (dyn core::ops::function::FnOnce<(), Output=()> + core::marker::Send), alloc::alloc::Global> (boxed.rs:1694)
==25281==    by 0x1D17F3: call_once<(), alloc::boxed::Box<(dyn core::ops::function::FnOnce<(), Output=()> + core::marker::Send), alloc::alloc::Global>> (unwind_safe.rs:271)
==25281==    by 0x1D17F3: do_call<core::panic::unwind_safe::AssertUnwindSafe<alloc::boxed::Box<(dyn core::ops::function::FnOnce<(), Output=()> + core::marker::Send), alloc::alloc::Global>>, ()> (panicking.rs:406)
==25281==    by 0x1D17F3: try<(), core::panic::unwind_safe::AssertUnwindSafe<alloc::boxed::Box<(dyn core::ops::function::FnOnce<(), Output=()> + core::marker::Send), alloc::alloc::Global>>> (panicking.rs:370)
==25281==    by 0x1D17F3: catch_unwind<core::panic::unwind_safe::AssertUnwindSafe<alloc::boxed::Box<(dyn core::ops::function::FnOnce<(), Output=()> + core::marker::Send), alloc::alloc::Global>>, ()> (panic.rs:133)
==25281==    by 0x1D17F3: run_test_in_process (lib.rs:608)
==25281==    by 0x1D17F3: test::run_test::run_test_inner::{{closure}} (lib.rs:500)
==25281==    by 0x19DA4D: {closure#1} (lib.rs:527)
==25281==    by 0x19DA4D: std::sys_common::backtrace::__rust_begin_short_backtrace (backtrace.rs:123)
==25281==    by 0x1A2857: {closure#0}<test::run_test::run_test_inner::{closure#1}, ()> (mod.rs:484)
==25281==    by 0x1A2857: call_once<(), std::thread::{impl#0}::spawn_unchecked::{closure#1}::{closure#0}> (unwind_safe.rs:271)
==25281==    by 0x1A2857: do_call<core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked::{closure#1}::{closure#0}>, ()> (panicking.rs:406)
==25281==    by 0x1A2857: try<(), core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked::{closure#1}::{closure#0}>> (panicking.rs:370)
==25281==    by 0x1A2857: catch_unwind<core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked::{closure#1}::{closure#0}>, ()> (panic.rs:133)
==25281==    by 0x1A2857: {closure#1}<test::run_test::run_test_inner::{closure#1}, ()> (mod.rs:483)
==25281==    by 0x1A2857: core::ops::function::FnOnce::call_once{{vtable-shim}} (function.rs:227)
==25281==    by 0x204442: call_once<(), dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global> (boxed.rs:1694)
==25281==    by 0x204442: call_once<(), alloc::boxed::Box<dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global>, alloc::alloc::Global> (boxed.rs:1694)
==25281==    by 0x204442: std::sys::unix::thread::Thread::new::thread_start (thread.rs:106)
==25281==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==25281== 
==25281== 
==25281== Process terminating with default action of signal 11 (SIGSEGV)
==25281==  Bad permissions for mapped region at address 0x0
==25281==    at 0x0: ???
==25281==    by 0x14AE62: lilv_instance_get_extension_data (inline_fns.rs:106)
==25281==    by 0x14AE62: lilv::instance::Instance::extension_data (instance.rs:115)
==25281==    by 0x16EFD5: livi::features::worker::maybe_get_worker_interface (worker.rs:153)
==25281==    by 0x17986A: livi::plugin::Plugin::instantiate (plugin.rs:103)
==25281==    by 0x15439D: livi::tests::test_all_plugins (lib.rs:338)
==25281==    by 0x17B4B9: livi::tests::test_all_plugins::{{closure}} (lib.rs:291)
==25281==    by 0x13995D: core::ops::function::FnOnce::call_once (function.rs:227)
==25281==    by 0x1D2D72: call_once<fn(), ()> (function.rs:227)
==25281==    by 0x1D2D72: test::__rust_begin_short_backtrace (lib.rs:585)
==25281==    by 0x1D17F3: call_once<(), (dyn core::ops::function::FnOnce<(), Output=()> + core::marker::Send), alloc::alloc::Global> (boxed.rs:1694)
==25281==    by 0x1D17F3: call_once<(), alloc::boxed::Box<(dyn core::ops::function::FnOnce<(), Output=()> + core::marker::Send), alloc::alloc::Global>> (unwind_safe.rs:271)
==25281==    by 0x1D17F3: do_call<core::panic::unwind_safe::AssertUnwindSafe<alloc::boxed::Box<(dyn core::ops::function::FnOnce<(), Output=()> + core::marker::Send), alloc::alloc::Global>>, ()> (panicking.rs:406)
==25281==    by 0x1D17F3: try<(), core::panic::unwind_safe::AssertUnwindSafe<alloc::boxed::Box<(dyn core::ops::function::FnOnce<(), Output=()> + core::marker::Send), alloc::alloc::Global>>> (panicking.rs:370)
==25281==    by 0x1D17F3: catch_unwind<core::panic::unwind_safe::AssertUnwindSafe<alloc::boxed::Box<(dyn core::ops::function::FnOnce<(), Output=()> + core::marker::Send), alloc::alloc::Global>>, ()> (panic.rs:133)
==25281==    by 0x1D17F3: run_test_in_process (lib.rs:608)
==25281==    by 0x1D17F3: test::run_test::run_test_inner::{{closure}} (lib.rs:500)
==25281==    by 0x19DA4D: {closure#1} (lib.rs:527)
==25281==    by 0x19DA4D: std::sys_common::backtrace::__rust_begin_short_backtrace (backtrace.rs:123)
==25281==    by 0x1A2857: {closure#0}<test::run_test::run_test_inner::{closure#1}, ()> (mod.rs:484)
==25281==    by 0x1A2857: call_once<(), std::thread::{impl#0}::spawn_unchecked::{closure#1}::{closure#0}> (unwind_safe.rs:271)
==25281==    by 0x1A2857: do_call<core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked::{closure#1}::{closure#0}>, ()> (panicking.rs:406)
==25281==    by 0x1A2857: try<(), core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked::{closure#1}::{closure#0}>> (panicking.rs:370)
==25281==    by 0x1A2857: catch_unwind<core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked::{closure#1}::{closure#0}>, ()> (panic.rs:133)
==25281==    by 0x1A2857: {closure#1}<test::run_test::run_test_inner::{closure#1}, ()> (mod.rs:483)
==25281==    by 0x1A2857: core::ops::function::FnOnce::call_once{{vtable-shim}} (function.rs:227)
==25281==    by 0x204442: call_once<(), dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global> (boxed.rs:1694)
==25281==    by 0x204442: call_once<(), alloc::boxed::Box<dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global>, alloc::alloc::Global> (boxed.rs:1694)
==25281==    by 0x204442: std::sys::unix::thread::Thread::new::thread_start (thread.rs:106)
==25281== 
==25281== HEAP SUMMARY:
==25281==     in use at exit: 40,082,657 bytes in 609,753 blocks
==25281==   total heap usage: 1,684,340 allocs, 1,074,587 frees, 104,765,496 bytes allocated
==25281== 
==25281== LEAK SUMMARY:
==25281==    definitely lost: 7,560 bytes in 158 blocks
==25281==    indirectly lost: 14,280 bytes in 447 blocks
==25281==      possibly lost: 18,583 bytes in 395 blocks
==25281==    still reachable: 40,042,234 bytes in 608,753 blocks
==25281==         suppressed: 0 bytes in 0 blocks
==25281== Rerun with --leak-check=full to see details of leaked memory
==25281== 
==25281== For lists of detected and suppressed errors, rerun with: -s
==25281== ERROR SUMMARY: 4 errors from 3 contexts (suppressed: 0 from 0)
Segmentation fault (core dumped)
```

wmedrano commented:
`==25281==    by 0x14AE62: lilv_instance_get_extension_data (inline_fns.rs:106)`

I'll put in a fix.

wmedrano commented:
Possible workaround: use `std::mem::transmute` to cast to `Option<_>` in https://github.com/wmedrano/lilv-sys-rs/blob/master/src/inline_fns.rs#L106

Casting example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a634e1418d134bbb651a02b64f28dd94

ctsexton commented on src/features/worker.rs:
```diff
@@ -0,0 +1,259 @@
+pub(crate) unsafe fn maybe_get_worker_interface(
+    instance: &mut lilv::instance::ActiveInstance,
+) -> Option<lv2_sys::LV2_Worker_Interface> {
+    // TODO: Remove below after
+    // https://github.com/poidl/lv2_raw/issues/4 is fixed.
+    let descriptor = instance.instance().descriptor().unwrap();
+    type ExtDataFn = extern "C" fn(uri: *const u8) -> *const c_void;
+    let extension_data: Option<ExtDataFn> = std::mem::transmute(descriptor.extension_data);
+    extension_data?;
+    // Delete up to here.
```
Thanks for the example - this seems to have fixed it.

wmedrano commented on src/plugin.rs:
```diff
@@ -346,6 +415,22 @@ impl Instance {
+    /// Grab the worker before you send the plugin
+    /// instance to the realtime thread and hold
+    /// onto it to perform work asynchronously.
+    pub fn get_worker(&mut self) -> Option<worker::Worker> {
```
`take_worker`

wmedrano commented on src/features/worker.rs:
```diff
@@ -0,0 +1,259 @@
+/// };
+///
+/// let mut worker_manager = livi::WorkerManager::default();
+/// if let Some(worker) = instance.get_worker() {
```
`take_worker`.

wmedrano commented on src/plugin.rs:
```diff
@@ -316,6 +374,17 @@ impl Instance {
         }
         self.inner.run(ports.sample_count);
+
+        let worker_interface = worker::maybe_get_worker_interface(&mut self.inner);
```
Can you store the result of line 103 in the struct instead of calling `worker::maybe_get_worker_interface` again? Asking because worker::maybe_get_worker_interface contains an allocation (through `CString::new`) somewhere down the stack. We could also tweak the code to take a preconstructed `CStr`, but storing the result may be more efficient anyways.

Then you would do `if let Some(interface) = self.worker_interface.as_mut() { ... }`

ctsexton replied:
Done, thank you I didn't realize there was an allocation happening there. Out of interest how do you find where allocations are happening other than digging through the code?

wmedrano replied:
I read the code with the help of rust-analyzer.

Seems like the crate [assert_no_alloc](https://docs.rs/assert_no_alloc/latest/assert_no_alloc/) can be used for testing allocations on the Rust side.

wmedrano commented on Cargo.toml:
```diff
@@ -17,9 +17,12 @@ lilv = "0.2"
 log = "0.4"
 lv2-sys = "2"
 lv2_raw = "0.2"
+ringbuf = "0.2.6"
```
One last bit, can you bump the version of the livi crate from `"0.3.12"` to `"0.3.13"`?
